v0.2

pgkirsch opened the pull request:
This pull request makes GPfit object oriented, with fit-type-specific sub-classes (e.g. `SoftmaxAffine`) derived from a `_Fit` parent class #100.

These classes have `print` and `plot` methods (`plot_surface` and `plot_slices` for 2D fits).
As part of this change this PR also includes the ability to save fits both to pickle and to a text file (related to #44).

The documentation also gets a facelift, a restructure, installation instructions (#69) and a tutorial page (#7)

Also closes #8, #33, #54, #104.

pgkirsch committed:
First pass at Fit class (#100)

pgkirsch committed:
All unit tests run except plot_fit and constraint_set (#100)

* Made evaluate() a static method
* t_classes --> t_evaluate

pgkirsch committed:
Working plot_fits method (#100)

pgkirsch committed:
plot_fit_2d (#100)

pgkirsch committed:
plot_slices for 2D fits (#100)

* Renamed plot_fit (--> plot) and plot_fit_2d (--> plot_surface)

pgkirsch committed:
Fit.save() method (#100)

Also closes #44

pgkirsch committed:
savetxt method (#100)

* Added better (and deterministic) unit tests to t_fit

pgkirsch committed:
__repr__ method to replace print_fit (#100)

* print(fit) instead of fit.print_fit()
* No longer prints the fit automatically (partially addresses #8)

pgkirsch committed:
constraint_set method of Fit creates FitConstraintSet (#100)

pgkirsch committed:
Remove old logo from docs (#93)

pgkirsch committed:
Prettier documentation, __version__ (#7)

Thanks to Akshay for the css

pgkirsch committed:
Installation instructions in docs (#69)

pgkirsch committed:
Acknowledgements on docs (#7)

pgkirsch committed:
Docs improvements (#7)

* Glossary -> reference
* Re-ordering
* Delete duplicate gpfit101

pgkirsch committed:
Replace fitdata with more meaningful attributes

pgkirsch committed:
Repair link to paper (#7)

pgkirsch commented:
@bqpd @whoburg I think this is 80% done, but I would like your feedback/advice on a couple things:

1. Should this be v1.0 or v0.2? It completes changes the primary API so by semantic versioning rules it should be a major release, but it also feels silly to release v1.0 when GPkit hasn't even technically released v1.0 yet.
2. As currently written, a user creates instances of the Fit subclasses, e.g. `f = ImplicitSoftmaxAffine(...)` but that's kind of a cumbersome UX. I was thinking the interface should be `f = Fit()` where the default type is ISMA. But `SoftmaxAffine` etc. all subclass `Fit` as the base class. Any thoughts on how to resolve this? My initial thought was to rename the current base class `Fit` to `BaseFit` (or similar) and then use that as the base class for `MaxAffine`, `SoftmaxAffine`, `ImplicitSoftmaxAffine`. `Fit` would then be the user-facing class.
3. Related: any thoughts on how to slice and dice `fit.py` -- it's currently pretty long and contains both the base class and the 3 sub-classes.

bqpd commented:
1. naw go for 1.0, that seems right!
2. having users create named fits seems fine -- they'll probably assign it to a variable called "fit" anyway
3. how about moving ImplicitSoftmaxAffine etc. to a different file?

pgkirsch commented:
1. Cool thanks!
2. It's just a lot to type... And in a practical context, a user generally just wants the best quality fit they can get. Is there a downside to using a `Fit` alias?
3. I think that's the play. But then it's like should each class have it's own module etc. Anyway probably no need to obsess over it too much, just wanted to see if you had any strong feelings.

pgkirsch committed:
Pretty output when verbosity=1 (#104)

And add type attribute to Fit

pgkirsch committed:
Add basic tutorial to docs (#7)

* Clean up (and rename) examples code and output files

pgkirsch committed:
_Fit now a private class (#100)

pgkirsch commented:
After doing a little more research, it looks like the best practice way of doing this would be an [Object Factory](https://realpython.com/factory-method-python/#factory-method-as-an-object-factory).

pgkirsch commented:
Ha! I just realized that one way to resolve this is to create a function (`fit`) that returns the appropriate object. So this overhaul may be backwards compatible after all...

pgkirsch committed:
fit convenience function (#100)

* Preserves the old API
* Updated tutorial accordingly

pgkirsch committed:
File restructure (#100)

* maths directory
* move fit classes to classes.py

pgkirsch committed:
Tighter formatting of repr

* Remove whitespace around *
* Replace ** with ^ in __repr__ (but keep ** in savetxt)
* Align + with = for isma

Closes #54

pgkirsch committed:
Error data corrected and clarified (#108)

whoburg commented:
@pgkirsch great work on the overhaul.

1. I'm good with v 1.0 as well.
2. One option is roughly matching scikit-learn (see e.g. https://scikit-learn.org/stable/modules/linear_model.html#lasso), where there are many types of `linear_model`, e.g. `linear_model.Lasso`, `linear_model.RidgeCV`, `linear_model.ElasticNet`. Yes these are long to type but I tend to like giving users full control. That said, scikit-learn is certainly not the only package to pattern after, and if you see good reasons to go another way I'm all for it.
3. No strong feelings on single module vs splitting up as long as we're not breaking the `pylint` module line count limit (1000 lines?)

pgkirsch commented:
Thanks @whoburg.

1. I'm torn on this. Now that the `fit` interface still exists (and is backwards compatible) it feels appropriate to v0.2. The only non-backwards compatible interface changes are the eliminated of `plot_fit_1d` and `print_fit` but I think I was probably the only person who used those.
2. What I think I've settled on is a best of both worlds that allows users to directly specify the class of fit they want to use (`MaxAffine`/`SoftmaxAffine`/`ImplicitSoftmaxAffine`) or just use the `fit` convenience function for the new and/or lazy user (see `fit.py` for this very short function). The new docs explains that both are options:
<img width="921" alt="Screen Shot 2021-07-28 at 10 31 07 PM" src="https://user-images.githubusercontent.com/8702579/127437307-20a3c5e2-599d-47af-b952-3f6b90e44176.png">
3. Ok. All are well below that length fortunately.

whoburg commented:
1. I'm also just fine with v0.2. No strong feelings from me on this.
2. That looks awesome. Love it.

pgkirsch committed:
pylint

pgkirsch commented:
Think I'm all done on this now! @bqpd @whoburg ready for review! (as much as you have the time or inclination to provide)

whoburg commented on docs/source/conf.py:
```diff
@@ -1,267 +1,72 @@
+#
+# import os
+# import sys
+# sys.path.insert(0, os.path.abspath('.'))
```
remove commented out code entirely?

pgkirsch replied:
Done

whoburg commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,65 @@
+
+Output:
+
+.. literalinclude:: examples/ex2_output.txt
```
this is fantastic documentation.

pgkirsch replied:
Thanks!

whoburg commented on gpfit/__init__.py:
```diff
@@ -1 +1,4 @@
 "Package for GP-compatible data fitting"
+__version__ = "1.0.0"
```
@pgkirsch mentioned he might want to change this back to 0.2? No preference from me, just flagging it to double check

pgkirsch replied:
Thanks for flagging with go through and make sure it's 0.2 everywhere.

whoburg commented on gpfit/classes.py:
```diff
@@ -1,135 +1,420 @@
+# pylint: disable=too-many-arguments
+# pylint: disable=too-many-branches
+# pylint: disable=import-error
+# pylint: disable=too-many-instance-attributes
```
it's a bit surprising that we need all these, but no worries if we really do

pgkirsch replied:
Good catch. I was able to remove a couple from here and a few more from `constraint_set.py`.

whoburg commented on gpfit/classes.py:
```diff
@@ -1,135 +1,420 @@
-"""
```
maybe call this module `fits.py` or `fit_classes.py`? Low priority no need to change

pgkirsch replied:
Yeah I'm not entirely satisfied with this module name either. Tried to address a little with the module docstring, but agree it's not ideal. `fits.py` seems confusingly close to `fit.py`. I like `fit_classes.py` but to me it feels like it still has the same ambiguity of "does it mean python class or class of fit?" and I erred on the side of the succinct. I think I'll sit with `classes.py` for now and if it still feels weird then readdress it in a future release. At least I didn't go full anarchy and call it `function_classes.py` haha.

whoburg commented on gpfit/tests/t_initialize.py:
```diff
@@ -3,7 +3,7 @@
 from numpy import arange, newaxis, vstack, log, exp
 from numpy.random import random_sample
-from gpfit.initialize import get_initial_parameters
+from gpfit.maths.initialize import get_initial_parameters
```
great job sneaking in the plural `maths` :) approved.

pgkirsch replied:
Haha thanks thought you might enjoy

whoburg reviewed (approved):
@pgkirsch it's so great to see this code get some love. Approved but only once tests are passing.

pgkirsch committed:
v0.2.0 (#105)

pgkirsch committed:
Remove unneeded pylint disables

pgkirsch committed:
Delete commented code in conf.py

* And update link in README

pgkirsch commented:
Thanks for the review @whoburg! Happy to finally be able to clean GPfit up a little. It looks like tests are only failing on Reynolds (as on master, not sure why) so I will see if Marshall or Berk are able to help us with this.

bqpd commented:
test this please

bqpd commented on docs/source/acknowledgements.rst:
```diff
@@ -0,0 +1,9 @@
+We thank the following people for helping to improve GPfit:
+
+* Marshall Galbraith for setting up continuous integration
+* `Akshay Agrawal`_ for the pretty docs css
```
ooh excited to see what it looks like

pgkirsch replied:
I like it a lot!

bqpd commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,65 @@
+Tutorial
```
love this, just laser-targeted at the primary use case

pgkirsch replied:
Thanks!

bqpd commented on docs/source/tutorial.rst:
```diff
@@ -0,0 +1,65 @@
+Tutorial
+********
+
+Given log-transformed data `x` and `y`, we can create a fit with `K` terms with
```
"log-transformed data" might stand some unpacking

pgkirsch replied:
Good call

bqpd commented:
the problem is reynolds is using python3.5..

bqpd commented:
@galbramc it looks like the python we get on reynolds with `virtualenv --python=python3 $WORKSPACE/venv2_gpfit` is python 3.5, and after sshing in it looks like that's the only one installed. Since that's about to be deprecated (and doesn't support @pgkirsch's fancy format strings), might this be the time to install a new one?

galbramc commented:
I can't easily install a new version of python. However, reynolds is running Ubuntu 16.04 and needs to be upgraded to 20.04. I was going to start working on that upgrade some time next week. That will give you python 3.8.

bqpd commented:
fabulous!

On Thu, Jul 29, 2021, 16:47 Marshall Galbraith ***@***.***>
wrote:

> I can't easily install a new version of python. However, reynolds is
> running Ubuntu 16.04 and needs to be upgraded to 20.04. I was going to
> start working on that upgrade some time next week. That will give you
> python 3.8.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/convexengineering/gpfit/pull/105#issuecomment-889444342>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALKAGB54WOO55JPVBDYCITT2G44VANCNFSM5BC33E2A>
> .
>

pgkirsch commented:
Thanks @galbramc!

pgkirsch committed:
Uniform docstrings using numpy style for autodoc (#7)

* Also includes _Fit documentation (its __init__ is important)
* And orders classes with same order as in source

pgkirsch committed:
fit.error (float) and fit.errors (dict) (#108)

* rename fit.error to fit.errors and add log error
* fit.error is a float (=fit.errors["rms_rel"])
* add a note explaining distinction in tutorial
* also fix the overloaded use of `f` in tutorial

pgkirsch committed:
More accurate gpfit101; log transform explained (#7)

pgkirsch commented:
@bqpd hopefully https://github.com/convexengineering/gpfit/pull/105/commits/7b41fb1da934ac4b0065227001f51da2d3b2b1e7 addresses your comment RE: unpacking log transform

bqpd commented:
yup that looks good!

pgkirsch commented:
@galbramc any chance you have time to upgrade Reynolds this week? I would also be totally fine with only using Macy's for GPfit tests for the time being, if that would be an easier change to make. Happy to make this change myself if I have permission and someone can point me in the right direction to get started.

galbramc commented:
I'm waiting for another computer we ordered to arrive before I head to campus to upgrade reynolds. I was hoping it would arrive last week, but it has still not shown up. So now I'm less certain when I'll be able to upgrade reynolds...

galbramc commented:
The unit tests are failing on macys as well though, so something is not right with this PR.

pgkirsch commented:
Ah ok thanks for the update @galbramc. And thanks for pointing that out -- I had stopped checking the reason for failure so missed that it was failing on Macys too. More impetus to fix #101.

pgkirsch committed:
Test coverage of different errors (#108)

pgkirsch committed:
Stricter check of error (#108)

pgkirsch committed:
Cleaner calculation of max errors (#108)

pgkirsch committed:
Include package data in setup (#101)

Create MANIFEST.in and include docs/ and test/artifacts/

galbramc commented:
test this please

galbramc commented:
reynolds has been upgraded. The testing errors are now legitimate errors.

pgkirsch commented:
Thank you @galbramc! 😭

galbramc commented:
test this please

pgkirsch committed:
Include gpfit.maths in setup.py (#101)

pgkirsch committed:
Updated build status icons in README

pgkirsch commented:
Huh pylint tests passing again, hurray. Hadn't made the pylintrc change yet.

galbramc commented:
test this please

galbramc commented:
test this please

galbramc commented:
The pylint success was a false positive. Jenkins was still not configured quite right, and there was a bug in the Jenkins script for running pyling. Note however that some pylint warnings are actually the master branch of gpfit.

pgkirsch commented:
Huh interesting that those warnings don't show up when I run pylint locally. Is Jenkins using a different (later?) version of pylint? (I'm apparently using 2.6.0). Where do I specify which version of pylint I want Jenkins to use given that it's not a package requirement?

galbramc commented:
The scripts create a virtualenv and installs the latest version of pylint. All the script does is runs the pylint.sh script in the gpfit repo otherwise.

pgkirsch committed:
Updated pylint; fixed the errors (#105)

pgkirsch commented:
Ah that was it -- updating to latest pylint (2.10) caught those 5 errors. Now fixed, hopefully the pylint tests pass on Jenkins now..

galbramc commented:
You got pylint fixed. Where are you expecting the artifacts directory to be located?

pgkirsch committed:
Change test to use assertLess (#105)

pgkirsch commented:
@galbramc sorry for the slow response. That week was hectic and I was on vacation this past week (nevertheless I should have let you know). Changed to assertLess as recommended.

galbramc commented:
test this please

galbramc commented:
test this please

galbramc commented:
ok we now get random failures on reynolds. take a look at the last two builds

galbramc commented:
test this please

galbramc commented:
test this please

pgkirsch commented:
Hm. It's nice to see the others passing at least! It seems like there must be some fundamental (maybe embarrassingly obvious) thing I don’t understand about how different OS’s do computation that would explain why these tests pass on Macys and Windows but not Reynolds. Does anything come to mind @galbramc?

As far as next step, the obvious (albeit lazy) thing to do here would be to relax the tolerance to `1e-2` for the `fsma` tests and call it good. Otherwise I'm not sure what other options exist.

galbramc commented:
I don't know anything about how the fsma tests work or what they are testing, so it's very hard for me to give advice here. I don't know if a tolerance of 1e-2 means you actually have a false positive, or if that is good enough. We need guidance from higher powers here (@whoburg)...

bqpd commented:
test this please

bqpd commented:
I've ssh'd in to reynolds and am running the build script, but haven't reproduced the error in four tries...

@galbramc are there any executables/environment variables that would be different in my user account compared to jenkins'? I haven't modified my bashrc at all and it's all but identical to jenkins':

```diff

40c40
<     xterm-color|*-256color) color_prompt=yes;;
---
>     xterm-color) color_prompt=yes;;
87,89d86
< # colored GCC warnings and errors
< #export GCC_COLORS='error=01;31:warning=01;35:note=01;36:caret=01;32:locus=01:quote=01'
< 
91c88
< alias ll='ls -alF'
---
> alias ll='ls -lhSr'
117a115,122
> 
> export PATH=$PATH:/jenkins/.local/bin
> export PATH=$PATH:/jenkins/bin
> 
> export pwid_LICENSE=5053@acdl.mit.edu
> 
> alias clang=/usr/bin/clang-8
> alias clang++=/usr/bin/clang++-8
```

bqpd commented:
wait I swear I can read trying again with those new PATHs

bqpd commented:
I've ssh'd in to reynolds and am running the build script, but haven't reproduced the error in four tries...

@galbramc are there any executables/environment variables that would be different in my user account compared to jenkins'? I haven't modified my bashrc at all and it's all but identical to jenkins'

EDIT: wait I swear I can read trying again with those new PATHs in that bashrc. deleted issue above bc it might have sensitive info

bqpd commented:
even with `source /home/jenkins/.bashrc` not getting any failures running `source $WORKSPACE/JenkinsGPkit/gpfit_unit_tests.sh` in `/home/eburn/gpfit` (with `WORKSPACE=/home/eburn/gpfit`)

bqpd commented:
test this please

(reynolds passed the previous round)

bqpd commented:
test this please

(reynolds passed the previous round)

galbramc commented:
The errors are bouncing around between machines. macys_VM failed the most recent one while reynolds passed. There is something in the tests that is not deterministic.

bqpd commented:
it looks like it's always the `ex2` example

bqpd commented:
can confirm nondeterminism just by running that example

bqpd commented:
in particular, the `SoftmaxAffine` fit varied from the mode in 3 of 10 trials, the others varied not at all

bqpd commented:
Regenerating u, x, y, k etc from the seed seems to fix this, implying that something in `*SoftmaxAffine` is modifying them unexpectedly

bqpd commented:
However, copying the values between calls and comparing them doesn't confirm that. @pgkirsch, not sure what's going on here, but you can kludge it by just putting

```python
np.random.seed(33404)
Vdd = np.random.random_sample(1000) + 1
Vth = 0.2*np.random.random_sample(1000) + 0.2
P = Vdd**2 + 30*Vdd*np.exp(-(Vth - 0.06*Vdd)/0.039)
u = np.vstack((Vdd, Vth))
w = P
x = np.log(u)
y = np.log(P)
K = 4
```

strangely this seems to vastly improve the consistency of the result but not in fact make it deterministic?????

bqpd commented:
Confirmed that this nondeterminism occurs even with a totally fixed x and y (attached, read with `x = np.genfromtxt("x.txt"); y = np.genfromtxt("y.txt")`)

[y.txt](https://github.com/convexengineering/gpfit/files/7146140/y.txt)
[x.txt](https://github.com/convexengineering/gpfit/files/7146141/x.txt)

galbramc commented:
So something in gpfit it self is non-deterministic. So is that a feature or a bug?

whoburg commented:
Shooting from the hip here, but [this article](https://towardsdatascience.com/stop-using-numpy-random-seed-581a9972805f) suggests that setting `np.random.seed` is bad practice, and instead "NumPy has moved towards advising users to create a random number generator for specific tasks or to even pass around when you need parts to be reproducible". Maybe worth a shot.

bqpd commented:
naw it's not random; as mentioned above a fixed input solves the problem,
and np.random isn't used elsewhere

On Fri, Sep 10, 2021, 17:01 Warren Hoburg ***@***.***> wrote:

> Shooting from the hip here, but this article
> <https://towardsdatascience.com/stop-using-numpy-random-seed-581a9972805f>
> suggests that setting np.random.seed is bad practice, and instead "NumPy
> has moved towards advising users to create a random number generator for
> specific tasks or to even pass around when you need parts to be
> reproducible". Maybe worth a shot.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/convexengineering/gpfit/pull/105#issuecomment-917296300>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALKAGAM4SSSAVUZQXE7OTDUBKL5NANCNFSM5BC33E2A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

pgkirsch commented:
Thanks for the extensive debugging @bqpd. But having read these in more detail now, I'm a little confused.

`np.random` _is_ actually used in `get_initial_parameters`, so using your fixed `x` and `y` values isn't enough to make the test deterministic (for any of MA, SMA or ISMA). But when I combine them with the fixed seed:
```python
np.random.seed(33404)                                                             
x = np.genfromtxt("x.txt"); y = np.genfromtxt("y.txt")                            
K = 4                                                                             
fma = MaxAffine(x, y, K, verbosity=1)                                             
fsma = SoftmaxAffine(x, y, K, verbosity=1)                                        
fisma = ImplicitSoftmaxAffine(x, y, K, verbosity=1)
```
I'm unable to reproduce the non-determinism on my laptop. I get the exact same fit for all 3 classes for 10 consecutive runs. And am I misunderstanding anything you did?

Note that all three fitting methods (MA, SMA, ISMA) use `get_initial_parameters` so any non-determinism should be true for all classes. I double checked and there is nothing I can see that would causes differences in determinism between SMA and the other two .

@whoburg's suggestion (which is actually already captured in #102 🙂 ) is maybe the only possible fix I can think of as well, but it's still unsatisfying given that I am not able to reproduce this problem locally.

(Also worth noting that this whole conversation is about a longstanding issue, #33)

bqpd commented:
ah my bad, searched for that other random call but didn't find it!

On Sat, Sep 11, 2021, 23:23 Philippe Kirschen ***@***.***>
wrote:

> Thanks for the extensive debugging @bqpd <https://github.com/bqpd>. But
> having read these in more detail now, I'm a little confused.
>
> np.random *is* actually used in get_initial_parameters, so using your
> fixed x and y values isn't enough to make the test deterministic (for any
> of MA, SMA or ISMA). But when I combine them with the fixed seed:
>
> np.random.seed(33404)
> x = np.genfromtxt("x.txt"); y = np.genfromtxt("y.txt")
> K = 4
> fma = MaxAffine(x, y, K, verbosity=1)
> fsma = SoftmaxAffine(x, y, K, verbosity=1)
> fisma = ImplicitSoftmaxAffine(x, y, K, verbosity=1)
>
> I'm unable to reproduce the non-determinism on my laptop. I get the exact
> same fit for all 3 classes for 10 consecutive runs. And am I
> misunderstanding anything you did?
>
> Note that all three fitting methods (MA, SMA, ISMA) use
> get_initial_parameters so any non-determinism should be true for all
> classes. I double checked and there is nothing I can see that would causes
> differences in determinism between SMA and the other two .
>
> @whoburg <https://github.com/whoburg>'s suggestion (which is actually
> already captured in #102
> <https://github.com/convexengineering/gpfit/issues/102> 🙂 ) is maybe the
> only possible fix I can think of as well, but it's still unsatisfying given
> that I am not able to reproduce this problem locally.
>
> (Also worth noting that this whole conversation is about a longstanding
> issue, #33 <https://github.com/convexengineering/gpfit/issues/33>)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/convexengineering/gpfit/pull/105#issuecomment-917566222>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AALKAGCUS7P33OBNJIRFG73UBRBNDANCNFSM5BC33E2A>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

whoburg commented:
Ha @pgkirsch nice job with #102 -- I wondered where I got the idea about random.seed being tricky -- it was from you!

pgkirsch committed:
Testing hard-code seed in initialize (#102)(#33)

pgkirsch committed:
Remove seed setting from TestFit functions (#102)(#33)

galbramc commented:
Yeah I'm guessing pylint got updated and they are now issuing new warnings. You need to decide if they should be fixed or suppressed. I can look at the Windoze issue.

galbramc commented:
test this please

galbramc commented:
test this please

pgkirsch committed:
Convert to f-string formatting (#105)

pgkirsch committed:
No more seeds anywhere (#105)

pgkirsch commented:
test this please

pgkirsch commented:
That's a good idea. Rather than running and seeing a value that fails can we just try setting to 0 (or either of the values from build 227/228) and running tests a few times? Would you mind doing that? I'm happy to do this but I don't think I have privileges.

pgkirsch commented:
test this please

bqpd commented:
ah darn. GPkit's had a bunch of python hash seed bugs, so I was hoping this might be one of those.

pgkirsch committed:
Old school debugging (#33)

pgkirsch committed:
More experimentation (#33)

pgkirsch commented:
test this please

pgkirsch commented:
test this please

pgkirsch commented:
test this please

pgkirsch commented:
test this please

pgkirsch commented:
test this please

pgkirsch commented:
test this please

galbramc commented:
So they reynolds failures indicate there is still some non-deterministic stuff going on in the tests you did not comment out.

galbramc commented:
Are the random numbers generated inside gpfit or are they generated as part of the unit tests?

pgkirsch committed:
Comment out more tests (#33)

pgkirsch committed:
See if time break being triggered

galbramc commented:
Yeah the timer kicks in and aborts the iterations then that would lead to non-deterministic behavior. The Jenkins nodes have different loads depending on other things running on them which means the required execution time always fluctuates.

pgkirsch committed:
pylint for the temporary code

pgkirsch commented:
test this please

pgkirsch commented:
Bingo:
<img width="1814" alt="Screen Shot 2021-10-17 at 10 26 29 PM" src="https://user-images.githubusercontent.com/8702579/137673857-28175275-9646-4967-b0f9-280bca3899c6.png">

pgkirsch committed:
Increase maxtime and un-comment tests (#33)

Max time was causing non-deterministic behaviour
Increasing to 20 seconds now (hopefully enough)

pgkirsch commented:
test this please

pgkirsch committed:
Max time infinite (#33)

pgkirsch commented:
test this please

pgkirsch commented:
test this please

pgkirsch committed:
Proper setup of random seed (#33)(#102)

* Remove hard-coded seed from get_initial_parameters
* Seed passed through functions to get_initial_parameters
* Default uses a None seed (random)
* Set fixed seed for unit tests

pgkirsch commented:
test this please

pgkirsch commented:
Looks like this is finally ready for a merge! @whoburg @bqpd  🎉 

Thank you @galbramc for all of your help and your patience with me!

bqpd commented:
@pgkirsch fantastic! congratulations!

whoburg commented:
WOW, incredible debugging effort @pgkirsch and @galbramc. Please accept my virtual standing ovation. 👏 👏 👏
